### PR TITLE
[AIRFLOW-510] Hide "paused" DAGs on the HomeView unless "Shown"

### PR DESCRIPTION
**Pull request by btallman**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- _https://issues.apache.org/jira/browse/AIRFLOW-510_

Testing Done:
- UI only, and as far as I can tell, we don't have UI test coverage yet

screen shot 2016-09-14 at 10 55 23 pm
screen shot 2016-09-14 at 10 55 36 pm

### CODE REVIEW

**Comment by jbalogh** [comment]
Shouldn't showing them be the default? I only go to the dag list page to pause and unpause dags. Hiding the paused dags means I have to learn where to toggle showing the paused dags before I can unpause it.

**Comment by btallman** [comment]
I'm not sure about your install, but in ours, the HomeView needs to be primarily a dashboard of current status. And there is no simple way to hide old DAGs. The default view for a developer is very different from that of a DevOps person.

**Comment by btallman** [comment]
I added screenshots to the PR, so people can see.

**Comment by jbalogh** [comment]
It sounds like you're using Pause as a proxy for hiding old dags? That seems like a somewhat orthogonal problem.

I use Pause as a temporary state when we need to make operational changes, so having the HomeView show paused dags is pretty important for understanding our current status.

**Comment by ldct** [comment]
I would prefer if show paused dags were the default (ie without any query params passed) and users could optionally hide paused dags, so that
1. The default is the same as what we have now
2. People who miss the "hide paused dags" link won't get confused if their dags disappear (they must opt-in to hiding dags)

File: airflow/www/templates/airflow/dags.html [hunk 1]
```diff
@@ -184,10 +189,11 @@
           $.ajax(url);
         });
       });
-      $('#dags').dataTable({
+      var dt = $('#dags').dataTable({
```
Comment (ldct) [inline airflow/www/templates/airflow/dags.html]: Use a more descriptive variable name

File: airflow/www/templates/airflow/dags.html [hunk 2]
```diff
@@ -184,10 +189,11 @@
...
         "iDisplayLength": 500,
         "bSort": false,
       });
+//      debugger;
```
Comment (ldct) [inline airflow/www/templates/airflow/dags.html]: what's this?

File: airflow/www/views.py [hunk 3]
```diff
@@ -1818,7 +1823,12 @@ def index(self):
         session.close()
 
         # get a list of all non-subdag dags visible to everyone
-        unfiltered_webserver_dags = [dag for dag in dagbag.dags.values() if not dag.parent_dag]
+        # optionally filter out "paused" dags
+        if request.args.get('showPaused') == 'True':
+            unfiltered_webserver_dags = [dag for dag in dagbag.dags.values() if not dag.parent_dag]
```
Comment (ldct) [inline airflow/www/views.py]: if we're adding yet another filter to the existing ones we need to rename the variables `do_filter`, `unfiltered_webserver_dags` etc

**Review by ldct — CHANGES_REQUESTED** [review]

**Comment by btallman** [comment]
That actually requires a config param, as the links would no longer solve
the underlying issue. The point is to have the default uncluttered so you
can scan the HomeView. Having to click a link to get to the Scanable View
makes not a useful feature. Basically, you only need to un-pause something
once, but you need to scan the list regularly.

Thanks,
Ben

## 

Ben Tallman - 503.680.5709

**Comment by ldct** [comment]
how about sorting the dags so that paused ones are displayed last? would that fit your use-case?

**Comment by btallman** [comment]
Well, not really.

Given the number of DAGs that get loaded into the DB over time, we need to
have a default hidden state for many of them. Pause works well for this, as
it means we are also hiding items that aren't accidentally running. Really,
we don't want them in the list, and our experience is that this is a
challenge for us, as we get into the hundreds of DAGs over time.

To be clear, we can run this as a cherry pick, but I can't really imagine
that this isn't a useful tool for most if not all users in a mature
installation.

Thanks,
Ben

## 

Ben Tallman - 503.680.5709

**Comment by btallman** [comment]
That is true Jeff, that's what I'm doing. How do you hide "old" dags?

Thanks,
Ben

## 

Ben Tallman - 503.680.5709

**Comment by jbalogh** [comment]
> That is true Jeff, that's what I'm doing. How do you hide "old" dags?

We don't have old dags. I think your team uses dags much differently than we do. Come to my meetup talk next week to hear more! 😉

**Comment by ldct** [comment]
> How do you hide "old" dags?

One way is to delete them :)

**Comment by btallman** [comment]
Yeah. That complicates historical lookup though... I agree that pause is orthogonal, but it actually solves or problem without adding a new field.

Get Outlook for iOS

On Thu, Sep 15, 2016 at 2:25 PM -0700, "Li Xuanji" notifications@github.com wrote:

How do you hide "old" dags?

One way is to delete them :)

**Comment by gwax** [comment]
It seems like it might be possible to get the best of all worlds via parameters in the airflow.cfg file. Perhaps default behavior stays show everything sorted alphabetical but with addition of a "Hide Paused" link.

Then there could be a config section for the UI (if there isn't already) with an option for `hide_paused` that determines whether they hide by default. And possibly even a `sort_paused_last` option as well.

**Comment by jbalogh** [comment]
Is it worth adding a `retired` knob to the dag model? I feel a little conflicted about that since it's almost the same thing as `paused`, but with a slightly more permanent intention.

I hope we can avoid adding more branches in config since that's hard for developers to work with and hard for users to understand.

**Comment by r39132** [comment]
@btallman 

I suggest adding 2 new configuration vars to airflow.cfg:
- `hide_paused_dags` (defaulted to False)
  - This determines whether the DAG Overview page hides DAGs by default. For your org, you may strongly consider setting it to True since you'd like to unclutter your UI.
- `enable_show_hide_paused_dags` (defaulted to False)
  - This determines whether your new feature is enabled/disabled. By default, it will be disabled and the expectation is that the DAG Overview page will look the same as it does today

At Agari, we sometimes pause a DAG and then forget to re-enable it. Hiding paused DAGs would make it easier for us to forget to re-enable them. However, I believe this feature is useful enough to allow merging given an additional configuration parameters.

Another thing I think you need to think about is that your paused dags are still getting parsed. The new scheduler launches a subprocess to parse each dag file.  You can look at `process_file` in `jobs.py`.

**Comment by ldct** [comment]
> That complicates historical lookup though...

What do you mean by historical lookup?

**Comment by r39132** [comment]
@btallman Any movement on this? I don't think that @zodiac's comment is a blocker. My proposal is that you can add your feature but hidden behind some configs.

**Comment by btallman** [comment]
Finishing the dag.backfill=False first, but will circle back on this
hopefully today.

Thanks,
Ben

## 

Ben Tallman - 503.680.5709

**Comment by r39132** [comment]
@btallman great. I will be on call starting next Tuesday and then traveling in Asia the week after, so unless you want to week 2+ weeks on me, best to get it in ASAP.

**Comment by r39132** [comment]
@btallman You have a bad merge/rebase here! You've recommitted 50+ previous commits. You'll need to rectify it before this can be considered.

**Comment by btallman** [comment]
Going to pull this one back, looking at the feedback, and streamline it based on a few of the comments.

**Pull request closed without merging**